feat(marketplace): integrate plugins-info plugin

ciiay opened the pull request:
## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

For [RHIDP-8301](https://issues.redhat.com/browse/RHIDP-8301)
This works with [rhdh pr #3366](https://github.com/redhat-developer/rhdh/pull/3366)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/redhat-developer/rhdh-plugins/blob/main/CONTRIBUTING.md#creating-changesets))
- [ ] Added or Updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

Test config:
```
dynamicPlugins:
  rootDirectory: dynamic-plugins-root
  frontend:
    default.main-menu-items:
      menuItems:
        ...
        default.admin:
          title: Administration
          icon: admin
...
    red-hat-developer-hub.backstage-plugin-marketplace:
      appIcons:
        - name: pluginsIcon
          importName: PluginsIcon
      dynamicRoutes:
        - path: /extensions
          importName: DynamicMarketplacePluginRouter
          menuItem:
            icon: pluginsIcon
            text: Extensions
      menuItems:
        extensions:
          parent: default.admin
```

Screen recording of testing on RHDH
Updated on 9/19 (with [rhdh pr #3366](https://github.com/redhat-developer/rhdh/pull/3366)) 

https://github.com/user-attachments/assets/63a3b166-3466-4871-a8b1-eec7687393e5

ciiay commented:
> Thank you @ciiay for this PR, looks great. I have some questions:
> 
> 1. For the empty state, can we use [this text](https://www.figma.com/design/GeVb7rc390TkxeQNnJIcBD/RHDH-Theme--pages-?node-id=1721-2069&t=7E1nkGjAjazIb3pC-1) and reduce the table height?
> 2. Can we move the Version column closer to the plugin name and the Actions column to the right-hand side. like [here](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=6479-4429&t=Kmymga3kWOKB8Ieq-1)?
> 3. I don't know if this is part of the PR but can we add tooltips that appear when hovering over the icons in the Actions column, like [here](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=6479-4429&t=Kmymga3kWOKB8Ieq-1)?

Thanks for the review. Uploaded the new screen recording for the fix.

ciiay commented on workspaces/marketplace/plugins/marketplace/src/components/InstalledPlugins/InstalledPluginsTable.tsx:
```diff
@@ -0,0 +1,192 @@
+                }}
+              >
+                <EditIcon />
+              </IconButton>
```
Do you mean the below option 1 or option 2?
Option 1:
<img width="3024" height="1352" alt="image" src="https://github.com/user-attachments/assets/35109b73-e934-4e54-9d5e-5d306c7def29" />

Option 2:
<img width="3024" height="1336" alt="image" src="https://github.com/user-attachments/assets/d779eb5e-802e-4045-91bf-f94cb1cb6504" />

Or maybe Option 3, something like this?
<img width="3024" height="1336" alt="image" src="https://github.com/user-attachments/assets/4a2ceebc-5bec-44f6-b197-bae970bea08a" />

If we don't plan to ship this before Actions column is ready, we can simply leave it as it is for now, right?

debsmita1 commented on workspaces/marketplace/plugins/marketplace/src/pages/DynamicMarketplacePluginRouter.tsx:
```diff
@@ -14,110 +14,101 @@
-// import { MarketplacePackagesTable } from '../components/MarketplacePackagesTable';
 import { MarketplacePackageDrawer } from '../components/MarketplacePackageDrawer';
 import { MarketplacePackageInstallPage } from './MarketplacePackageInstallPage';
+import { InstallationContextProvider } from '../components/InstallationContext';
```
Do you think we should rename this to "InstalledPluginsContextProvider" ?

ShiranHi commented:
> @ShiranHi the backend and frontend components will have different versions, how do we display that ?

Yes, discussed [here](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s?node-id=6479-4429#1418183561).

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/components/InstalledPlugins/InstalledPluginsTable.tsx:
```diff
@@ -0,0 +1,216 @@
+  const { count } = useInstalledPluginsCount();
+  const dynamicPluginInfo = useApi(dynamicPluginsInfoApiRef);
+  let data: DynamicPluginInfo[] = [];
+  const columns: TableColumn<DynamicPluginInfo>[] = [
```
This columns definition should be "static" (move it before the component) otherwise the installed plugins tab creates this warning:

```
The columns provided to material table are static, but contain functions which update on every render, resetting the table state. Provide a stable function or column reference or an row id to prevent state loss.
```

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/components/InstalledPlugins/InstalledPluginsTable.tsx:
```diff
@@ -0,0 +1,216 @@
+    setCurrentSearchTerm(search);
+
+    try {
+      // for now sorting/searching/pagination is handled client-side
+      const installedPlugins = await dynamicPluginInfo.listLoadedPlugins();
```
I know you just moved the code. But actual this code makes an api call every time the user changes the filter. But we can address this in a follow up PR.

Can you create a bug for this?

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/plugin.ts:
```diff
@@ -91,11 +107,12 @@ export const MarketplaceTabbedPageRouter = marketplacePlugin.provide(
 export const DynamicMarketplacePluginRouter = marketplacePlugin.provide(
   createRoutableExtension({
-    name: 'DynamicMarketplacePluginRouter',
+    name: 'MarketplaceRouter',
```
Why changing this name? As long as the export is called DynamicMarketplacePluginRouter and the component, I would keep the old name.

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/plugin.ts:
```diff
@@ -122,19 +139,9 @@ export const DynamicMarketplacePluginContent = marketplacePlugin.provide(
 /**
  * @public
  */
-export const InstallationContextProvider = marketplacePlugin.provide(
```
I know we discussed backward compatible today and I said I wouldn't be too much worried about it.

But can we export a null component here maybe? It was an old idea from @debsmita1 and it doesn't cost anything ;)

```tsx
/**
 * @public
 * @deprecated
 */
export const InstallationContextProvider = marketplacePlugin.provide(
  createComponentExtension({
    name: 'InstallationContextProvider',
    component: {
      sync: () => null,
    },
  }),
);
```

ciiay commented on workspaces/marketplace/plugins/marketplace/src/components/InstalledPlugins/InstalledPluginsTable.tsx:
```diff
@@ -0,0 +1,216 @@
+  ): Promise<QueryResult<DynamicPluginInfo>> => {
+    const {
+      orderBy = { field: 'name' },
+      orderDirection = 'asc',
```
Removed.

ciiay commented on workspaces/marketplace/.changeset/unlucky-hornets-cross.md:
```diff
@@ -0,0 +1,15 @@
+---
+'@red-hat-developer-hub/backstage-plugin-marketplace-backend': minor
+'@red-hat-developer-hub/backstage-plugin-marketplace': major
```
Hi @christoph-jerolimov , any idea here? Here's the gpt-5 analysis about this `major` change. Not sure if it's applicable to our project as well.

- Yes, setting the frontend plugin to major is correct. The null export of InstallationContextProvider is a breaking change for consumers importing it.
- Backend as minor is also appropriate given added functionality (plugins-info integration, new tab) without breaking its public API.

ciiay commented:
> Why is the "Install" disabled, can we provide a tooltip ?

Changed it to `Save` to align up with the plugin install page.

> Should be packages

Updated to `packages`, also updated the plugin install page to use `plugins` too so both are consistent.

debsmita1 reviewed (approved):
Verified that changes made to a package config, updates the yaml file configured in the extensions configuration, and the same is reflected in the edit plugin configuration as well

https://github.com/user-attachments/assets/c1bd0d4e-6717-4d56-926b-0803f40d2599

- When updating the package config, users should be directed to the "Installed Packages" tab instead of the package info drawer.
- Since there can be multiple packages with config updates, @ShiranHi could you guide us on how we should handle an alert message in this case?
- Also, we should hide the "Edit Instructions" section if no instructions are available for the package.
- @ShiranHi do you agree that we show the same "Actions" dropdown here like we have on the plugins info drawer (We can address this as part of https://issues.redhat.com/browse/RHIDP-6810)

<img width="1859" height="854" alt="Image" src="https://github.com/user-attachments/assets/bfc49004-ac20-485b-abc4-c5d69823546e" />

<img width="1859" height="854" alt="Image" src="https://github.com/user-attachments/assets/e30dd5d4-61b5-44a0-af6f-f151990b09f6" />

Let's address the issues in a separate PR 


/lgtm

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace-backend/package.json:
```diff
@@ -44,7 +45,6 @@
     "@red-hat-developer-hub/backstage-plugin-marketplace-common": "workspace:^",
     "express": "^4.17.1",
     "express-promise-router": "^4.1.0",
-    "yaml": "^2.7.1",
```
`plugins/marketplace-backend/src/installation/FileInstallationStorage.ts` is using the yaml package. Please move this back from devDependencies to dependencies.

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/components/InstalledPackages/InstalledPackagesTable.tsx:
```diff
@@ -0,0 +1,365 @@
+    ],
+    [location.pathname, location.search],
+  );
+  const fetchData = async (
+    query: Query<InstalledPackageRow>,
+  ): Promise<QueryResult<InstalledPackageRow>> => {
+    const { page = 0, pageSize = 5 } = query || {};
+
+    try {
+      // Ensure data available; avoid re-fetching on search or pagination
+      const installed = installedQuery.data ?? [];
+      const packagesResponse = packagesQuery.data ?? { items: [] as any[] };
+
+      const entitiesByName = new Map(
+        (packagesResponse.items ?? []).map(entity => [
+          (entity.metadata?.name ?? '').toLowerCase(),
+          entity,
+        ]),
+      );
+
+      // Build rows for ALL installed items; if no matching entity, mark missing
+      const rows: InstalledPackageRow[] = installed.map(p => {
+        const normalized = p.name
+          .replace(/[@/]/g, '-')
+          .replace(/-dynamic$/, '')
+          .replace(/^-+/, '')
+          .toLowerCase();
+        const entity = entitiesByName.get(normalized) as any | undefined;
+        const rawName = entity
+          ? (entity.metadata?.title as string) ||
+            (entity.metadata?.name as string)
+          : getReadableName(p.name);
+        const cleanedName = rawName.replace(/\s+(frontend|backend)$/i, '');
```
I know this is copied but this looks like a lot of code to me that the backstage table is already doing? But maybe I'm wrong in the detail.

Fine for now.

christoph-jerolimov reviewed (commented):
I have a bunch of comment but I think we can merge it now and then take a look into this together:

debsmita1 commented:
I will address these review comments as part of https://issues.redhat.com/browse/RHIDP-6810
